ctapipe/reco/HillasReconstructor.py

```diff
@@ -212,46 +211,77 @@ def initialize_hillas_planes(
             dictionary of pointing direction per each telescope
         array_pointing: SkyCoord[AltAz]
             pointing direction of the array
+
+        Notes
+        -----
+        The part of the algorithm taking into account divergent pointing
+        mode and the correction to the psi angle is explained in [gasparetto]_
+        section 7.1.4.
         """
 
         self.hillas_planes = {}
         k = next(iter(telescopes_pointings))
         horizon_frame = telescopes_pointings[k].frame
         for tel_id, moments in hillas_dict.items():
-            # we just need any point on the main shower axis a bit away from the cog
-            p2_x = moments.x + 0.1 * u.m * np.cos(moments.psi)
-            p2_y = moments.y + 0.1 * u.m * np.sin(moments.psi)
-            focal_length = subarray.tel[tel_id].optics.equivalent_focal_length
 
             pointing = SkyCoord(
                 alt=telescopes_pointings[tel_id].alt,
                 az=telescopes_pointings[tel_id].az,
                 frame=horizon_frame,
             )
 
-            camera_frame = CameraFrame(
-                focal_length=focal_length, telescope_pointing=pointing
-            )
+            focal_length = subarray.tel[tel_id].optics.equivalent_focal_length
 
-            cog_coord = SkyCoord(x=moments.x, y=moments.y, frame=camera_frame,)
-            cog_coord = cog_coord.transform_to(horizon_frame)
+            if moments.x.unit == u.m:  # Image parameters are in CameraFrame
 
-            p2_coord = SkyCoord(x=p2_x, y=p2_y, frame=camera_frame)
-            p2_coord = p2_coord.transform_to(horizon_frame)
+                # we just need any point on the main shower axis a bit away from the cog
+                p2_x = moments.x + 0.1 * u.m * np.cos(moments.psi)
+                p2_y = moments.y + 0.1 * u.m * np.sin(moments.psi)
 
-            # re-project from sky to a "fake"-parallel-pointing telescope
-            # then recalculate the psi angle
-            if self.divergent_mode:
-                camera_frame_parallel = CameraFrame(
-                    focal_length=focal_length, telescope_pointing=array_pointing
+                camera_frame = CameraFrame(
+                    focal_length=focal_length, telescope_pointing=pointing
+                )
```

Comment:
we should check if this is a very slow operation: creating a Frame for each event rather than creating it once, and updating its frame attributes like the pointing.  Could have a speed impact.  Though I seem to recall updating frame attributes was not trivial in astropy.

Reply:
I changed my mind: let's ignore this for now. I have some ideas for speeding up the reconstruction overall that can go into another PR.